caas/kubernetes/provider/rbac.go

```diff
@@ -357,44 +362,84 @@ func (k *kubernetesClient) listRoles(selector k8slabels.Selector) ([]rbacv1.Role
+// createClusterRole attempts to create the provided cluster role in the
+// cluster. If the role already exists a Juju AlreadyExist error is returned
+// along with the the fetched objected from the cluster.
+func createClusterRole(
+	api rbacapi.ClusterRoleInterface,
+	cRole *rbacv1.ClusterRole,
+) (*rbacv1.ClusterRole, error) {
 	utils.PurifyResource(cRole)
-	out, err := k.client().RbacV1().ClusterRoles().Create(context.TODO(), cRole, v1.CreateOptions{})
+	out, err := api.Create(context.TODO(), cRole, v1.CreateOptions{})
 	if k8serrors.IsAlreadyExists(err) {
 		return nil, errors.AlreadyExistsf("cluster role %q", cRole.GetName())
 	}
-	return out, errors.Trace(err)
+	return out, errors.Annotatef(err, "creating cluster role %q", cRole.GetName())
 }
 
-func (k *kubernetesClient) updateClusterRole(cRole *rbacv1.ClusterRole) (*rbacv1.ClusterRole, error) {
-	out, err := k.client().RbacV1().ClusterRoles().Update(context.TODO(), cRole, v1.UpdateOptions{})
+// updateClusterRole attempts to update the provided ClusterRole. If the object
+// doesn't already exist inside of the Kubernetes cluster then a Juju not found
+// error is returned.
+func updateClusterRole(
+	api rbacapi.ClusterRoleInterface,
+	cRole *rbacv1.ClusterRole,
+) (*rbacv1.ClusterRole, error) {
+	out, err := api.Update(context.TODO(), cRole, v1.UpdateOptions{})
 	if k8serrors.IsNotFound(err) {
 		return nil, errors.NotFoundf("cluster role %q", cRole.GetName())
 	}
 	return out, errors.Trace(err)
 }
 
-func (k *kubernetesClient) ensureClusterRole(cRole *rbacv1.ClusterRole) (out *rbacv1.ClusterRole, cleanups []func(), err error) {
-	out, err = k.createClusterRole(cRole)
+func ensureClusterRole(
+	api rbacapi.ClusterRoleInterface,
+	cRole *rbacv1.ClusterRole,
+	claims ...utils.Claim,
+) (out *rbacv1.ClusterRole, cleanups []func(), err error) {
+	out, err = createClusterRole(api, cRole)
 	if err == nil {
 		logger.Debugf("cluster role %q created", out.GetName())
-		cleanups = append(cleanups, func() { _ = k.deleteClusterRole(out.GetName(), out.GetUID()) })
+		cleanups = append(cleanups, func() { _ = deleteClusterRole(api, out.GetName(), out.GetUID()) })
 		return out, cleanups, nil
 	}
+
 	if !errors.IsAlreadyExists(err) {
-		return nil, cleanups, errors.Trace(err)
+		return nil, cleanups, errors.Annotatef(
+			err,
+			"ensuring cluster role %s",
+			cRole.Name,
+		)
 	}
-	_, err = k.listClusterRoles(utils.LabelsToSelector(cRole.GetLabels()))
+
+	existing, err := api.Get(context.TODO(), cRole.Name, v1.GetOptions{})
 	if err != nil {
-		if errors.IsNotFound(err) {
-			// cRole.Name is already used for an existing cluster role.
-			return nil, cleanups, errors.AlreadyExistsf("cluster role %q", cRole.GetName())
+		return nil, cleanups, errors.Annotatef(
+			err,
+			"ensuring cluster role %s",
+			cRole.Name,
+		)
+	}
+
+	if len(claims) != 0 {
+		hasClaim, err := utils.ClaimAggregateOr(claims...).Assert(existing)
```

Comment:
We could just drop the `len(claims)` check and have `ClaimAggregator().Assert()` return `true` if there are no claims. This would make it cleaner here.